<a href="https://colab.research.google.com/github/joeyMckinney/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/Josiah_McKinney_LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [3]:
### YOUR WORK HERE
import numpy as np
import pandas as pd
from scipy import stats

column_headers = ['symboling', 'normalized_losses', 'make', 'fuel-type', 
                  'aspiration', 'num_of_doors', 'body_style', 'drive_wheels', 
                  'engine_location', 'wheel_base', 'length', 'width', 'height', 
                  'curb-weight', 'engine_type', 'num_of_cylinders', 
                  'engine_size', 'fuel_system', 'bore', 'stroke', 
                  'compression_ratio', 'horsepower', 'peak_rpm', 'city_mpg', 
                  'highway_mpg', 'price']
df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data', names=column_headers, na_values='?')


In [4]:
df.head()

symboling  normalized_losses         make  ... city_mpg highway_mpg    price
0          3                NaN  alfa-romero  ...       21          27  13495.0
1          3                NaN  alfa-romero  ...       21          27  16500.0
2          1                NaN  alfa-romero  ...       19          26  16500.0
3          2              164.0         audi  ...       24          30  13950.0
4          2              164.0         audi  ...       18          22  17450.0

[5 rows x 26 columns]

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and compare your sample mean estimate to that value. Can you say that they are different? 

In [8]:
### YOUR WORK HERE
sample_cars = df.sample(10, random_state=30)
sample_cars.head(10)

symboling  normalized_losses  ... highway_mpg    price
121          1              154.0  ...          38   6692.0
176         -1               65.0  ...          32  10898.0
24           1              148.0  ...          38   6229.0
88          -1              137.0  ...          30   9279.0
72           3              142.0  ...          18  35056.0
103          0              108.0  ...          25  13499.0
26           1              148.0  ...          38   7609.0
20           0               81.0  ...          43   6575.0
155          0               91.0  ...          32   8778.0
185          2               94.0  ...          34   8195.0

[10 rows x 26 columns]

In [10]:
print(df['highway_mpg'].mean())
print(sample_cars['highway_mpg'].mean())
print('yes they are different')

30.75121951219512
32.8
yes they are different


In [11]:
stats.ttest_1samp(sample_cars['highway_mpg'], df['highway_mpg'].mean())

Ttest_1sampResult(statistic=0.8912422391619923, pvalue=0.39600040771093326)

based on a t-statistic of .8912, and a p-vlaue of .396 > .05, i fail reject the null hypothesis that the sample mean and the population mean are the same.






---



The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

In [12]:
### YOUR WORK HERE
stats.ttest_1samp(sample_cars['highway_mpg'], 35)

Ttest_1sampResult(statistic=-0.9570244044334747, pvalue=0.36355472283248624)

based on a t-statistic of .8912, and a p-vlaue of .396 > .05, i fail to reject the null hypothesis that the sample mean and the population mean are the same.

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, what would you conclude when sampling 100 cars?

In [13]:
### YOUR WORK HERE
sample_cars_100 = df.sample(100, random_state=30)
sample_cars_100.shape

(100, 26)

In [14]:
stats.ttest_1samp(sample_cars_100['highway_mpg'], 35)

Ttest_1sampResult(statistic=-6.5185242467032305, pvalue=3.003137417468754e-09)

based on a t-statistic of -0.166, and a p-vlaue of approx .0, i reject the null hypothesis that the sample mean and the population mean are the same.

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

because the smaller the sample size the more liky change that you are wrong. and by wrong i have a less change of representing the whole population.

## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [15]:
### YOUR WORK HERE
cols = [
        'Party',
        'handicapped-infants',
        'water-project-cost-sharing',
        'adoption-of-the-budget-resolution',
        'physician-fee-freeze',
        'el-salvador-aid',
        'religious-groups-in-schools',
        'anti-satellite-test-ban',
        'aid-to-nicaraguan-contras',
        'mx-missile',
        'immigration',
        'synfuels-corporation-cutback',
        'education-spending',
        'superfund-right-to-sue',
        'crime',
        'duty-free-exports',
        'export-administration-act-south-africa',
        ]
df2 = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data', names=cols, na_values='?')
df2.head()


Party  ... export-administration-act-south-africa
0  republican  ...                                      y
1  republican  ...                                    NaN
2    democrat  ...                                      n
3    democrat  ...                                      y
4    democrat  ...                                      y

[5 rows x 17 columns]

In [16]:
df2 = df2.replace({'y' : 1, 'n' : 0})
df2.head()

Party  ...  export-administration-act-south-africa
0  republican  ...                                     1.0
1  republican  ...                                     NaN
2    democrat  ...                                     0.0
3    democrat  ...                                     1.0
4    democrat  ...                                     1.0

[5 rows x 17 columns]

In [17]:
rep = df2[df2['Party']=='republican']
dem = df2[df2['Party']=='democrat']
print(rep.shape)
rep.head()

(168, 17)


Party  ...  export-administration-act-south-africa
0   republican  ...                                     1.0
1   republican  ...                                     NaN
7   republican  ...                                     1.0
8   republican  ...                                     1.0
10  republican  ...                                     0.0

[5 rows x 17 columns]

In [18]:
print(dem.shape)
dem.head()

(267, 17)


Party  ...  export-administration-act-south-africa
2  democrat  ...                                     0.0
3  democrat  ...                                     1.0
4  democrat  ...                                     1.0
5  democrat  ...                                     1.0
6  democrat  ...                                     1.0

[5 rows x 17 columns]

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

In [19]:
print(dem.isnull().sum())
rep.isnull().sum()

Party                                      0
handicapped-infants                        9
water-project-cost-sharing                28
adoption-of-the-budget-resolution          7
physician-fee-freeze                       8
el-salvador-aid                           12
religious-groups-in-schools                9
anti-satellite-test-ban                    8
aid-to-nicaraguan-contras                  4
mx-missile                                19
immigration                                4
synfuels-corporation-cutback              12
education-spending                        18
superfund-right-to-sue                    15
crime                                     10
duty-free-exports                         16
export-administration-act-south-africa    82
dtype: int64


Party                                      0
handicapped-infants                        3
water-project-cost-sharing                20
adoption-of-the-budget-resolution          4
physician-fee-freeze                       3
el-salvador-aid                            3
religious-groups-in-schools                2
anti-satellite-test-ban                    6
aid-to-nicaraguan-contras                 11
mx-missile                                 3
immigration                                3
synfuels-corporation-cutback               9
education-spending                        13
superfund-right-to-sue                    10
crime                                      7
duty-free-exports                         12
export-administration-act-south-africa    22
dtype: int64

In [20]:
dem.mean(axis=0)


handicapped-infants                       0.604651
water-project-cost-sharing                0.502092
adoption-of-the-budget-resolution         0.888462
physician-fee-freeze                      0.054054
el-salvador-aid                           0.215686
religious-groups-in-schools               0.476744
anti-satellite-test-ban                   0.772201
aid-to-nicaraguan-contras                 0.828897
mx-missile                                0.758065
immigration                               0.471483
synfuels-corporation-cutback              0.505882
education-spending                        0.144578
superfund-right-to-sue                    0.289683
crime                                     0.350195
duty-free-exports                         0.637450
export-administration-act-south-africa    0.935135
dtype: float64

In [21]:
rep.mean(axis=0)

handicapped-infants                       0.187879
water-project-cost-sharing                0.506757
adoption-of-the-budget-resolution         0.134146
physician-fee-freeze                      0.987879
el-salvador-aid                           0.951515
religious-groups-in-schools               0.897590
anti-satellite-test-ban                   0.240741
aid-to-nicaraguan-contras                 0.152866
mx-missile                                0.115152
immigration                               0.557576
synfuels-corporation-cutback              0.132075
education-spending                        0.870968
superfund-right-to-sue                    0.860759
crime                                     0.981366
duty-free-exports                         0.089744
export-administration-act-south-africa    0.657534
dtype: float64

In [22]:
##this is an issue where dems support the bill more
dem['adoption-of-the-budget-resolution'].mean()

0.8884615384615384

In [23]:
rep['adoption-of-the-budget-resolution'].mean()

0.13414634146341464

1)null hypothesis:

dems support this bill at the same level that republics do

$H_0: \overline{x}_r == \overline{x}_d$



2)alternative hypothesis:
dems support this bill in a different proportion than reps

$H_a: \overline{x}_r \neq \overline{x}_d$


3)confidence level 99%

In [24]:
stats.ttest_ind(dem['adoption-of-the-budget-resolution'],rep['adoption-of-the-budget-resolution'], nan_policy='omit')

Ttest_indResult(statistic=23.21277691701378, pvalue=2.0703402795404463e-77)

4)conclusion:

due to a t-statistic of 23.31 and p-value of ~0 we reject the null hypothesis that democrats and republics support the adoption of the bughet resolution bill at the same rate, and suggest the alternative that the dems supports the bill at a higher rate.

In [25]:
##this is an issue where reps support the bill more
dem['crime'].mean()

0.35019455252918286

In [26]:
rep['crime'].mean()

0.9813664596273292

1)null hypothesis:

dems support this bill at the same level that republics do

$H_0: \overline{x}_r == \overline{x}_d$



2)alternative hypothesis:

reps support this bill in a different proportion than dems

$H_a: \overline{x}_r \neq \overline{x}_d$


3)confidence level 99%

In [27]:
stats.ttest_ind(dem['crime'],rep['crime'], nan_policy='omit')

Ttest_indResult(statistic=-16.342085656197696, pvalue=9.952342705606092e-47)

4)conclusion:

due to a t-statistic of -16.34 and p-value of ~0 we reject the null hypothesis that democrats and republics support the crime bill at the same rate, and suggest the alternative that the reps supports the bill at a higher rate.


In [28]:
##this is an issue where reps and dems suppor the same
dem['water-project-cost-sharing'].mean()

0.502092050209205

In [29]:
rep['water-project-cost-sharing'].mean()

0.5067567567567568

1)null hypothesis:

dems support this bill at the same level that republics do

$H_0: \overline{x}_r == \overline{x}_d$



2)alternative hypothesis:

reps support this bill in a different proportion than dems

$H_a: \overline{x}_r \neq \overline{x}_d$


3)confidence level 90%

In [30]:
stats.ttest_ind(rep['water-project-cost-sharing'],dem['water-project-cost-sharing'], nan_policy='omit')

Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)

4)conclusion:

due to a t-statistic of 0.088 and p-value of 0.929 we fail to reject the null hypothesis that democrats and republics support the crime bill at the same rate.

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [31]:
print(rep.head())
dem.head()

         Party  ...  export-administration-act-south-africa
0   republican  ...                                     1.0
1   republican  ...                                     NaN
7   republican  ...                                     1.0
8   republican  ...                                     1.0
10  republican  ...                                     0.0

[5 rows x 17 columns]


Party  ...  export-administration-act-south-africa
2  democrat  ...                                     0.0
3  democrat  ...                                     1.0
4  democrat  ...                                     1.0
5  democrat  ...                                     1.0
6  democrat  ...                                     1.0

[5 rows x 17 columns]

In [32]:
stats.ttest_ind(rep['handicapped-infants'], dem['handicapped-infants'], nan_policy='omit')

Ttest_indResult(statistic=-9.205264294809222, pvalue=1.613440327937243e-18)

In [33]:
### YOUR WORK HERE
#a fuction that uses ttest_ind to find pvalues less than 0 to let us know if it fails to reject or rejects

def t_test(df1, df2, list_of_col_1, list_of_col_2):
  tests = []
  for i in list_of_col_1:
    col_1 = df1[i]
    col_2 = df2[i]
    t = stats.ttest_ind(col_1,col_2, nan_policy='omit')
    tests.append(t)
  return tests

def fail_to_reject_or_not(df1, df2):
  df1_cols = []
  df2_cols = []
  final_results = []
  for i in df1.columns:
    df1_cols.append(i)
  for i in df2.columns:
    df2_cols.append(i)
  df1_cols.pop(0)
  df2_cols.pop(0)
  results = t_test(df1, df2, df1_cols, df2_cols)
  for i in results:
    if i.pvalue < .01:
      final_results.append('i reject')
    elif i.pvalue > .01:
      final_results.append('i fail to reject')
    else:
      final_results.append('somethings went wrong')
  a = list(dem.columns.values)
  a.pop(0)
  d = {'bill': a, 'results': final_results}
  final = pd.DataFrame(data=d)
  return final

In [34]:
fail_to_reject_or_not(rep, dem)

bill           results
0                      handicapped-infants          i reject
1               water-project-cost-sharing  i fail to reject
2        adoption-of-the-budget-resolution          i reject
3                     physician-fee-freeze          i reject
4                          el-salvador-aid          i reject
5              religious-groups-in-schools          i reject
6                  anti-satellite-test-ban          i reject
7                aid-to-nicaraguan-contras          i reject
8                               mx-missile          i reject
9                              immigration  i fail to reject
10            synfuels-corporation-cutback          i reject
11                      education-spending          i reject
12                  superfund-right-to-sue          i reject
13                                   crime          i reject
14                       duty-free-exports          i reject
15  export-administration-act-south-africa          i reject



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





In [ ]:
### YOUR WORK HERE

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [ ]:
### YOUR WORK HERE